proto-lens/src/Data/ProtoLens/Compare.hs

```diff
@@ -0,0 +1,108 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+-- | Matchers for comparing protocol buffer messages.
+module Data.ProtoLens.Compare (
+    -- * Matchers
+    partial,
+    ) where
+
+import Data.Constraint (Dict (..))
+import qualified Data.Map as Map
+import Lens.Family2 (Lens', over, set, view)
+import Lens.Family2.Unchecked (lens)
+import Data.ProtoLens.Message
+
+-- | Perform a partial match on a protobuffer. This corresponds to equality for
+-- primitive fields, (recursive) partial equality for required 'Message'
```

Comment:
We should probably be more explicit in the comment: this is equality for proto3 scalar optional fields and proto2 required fields.